src/asmcnc/comms/grbl_settings_manager.py

```diff
@@ -250,3 +255,32 @@ def on_persistent_setting(self, instance, value, setting):
             # restore default value:
             self.machine.write_dollar_setting(setting, self._system_default_data[setting])
             log('Restored {} to default: {}'.format(descriptions[setting], self._system_default_data[setting]))
+            
+    def on_console_specific_setting(self, instance, value, setting):
+        """
+        Called when a console specific setting is read in or changed.
+
+        Checks if the read in value differs from last recorded value.
+        """
+        descriptions = {
+            51: "SC2 ($51) setting"
+        }
+
+        # Check if this is the first read
+        if not self._has_read_51:
+            # If setting has not yet been saved then save it
+            if self._machine_saved_data[setting] == -1:
+                self._machine_saved_data[setting] = value
+                self.save_machine_data_to_file()
+                log('First startup after update? Setting {} saved to file: {}'.format(descriptions[setting], value))
+            # Otherwise if the value does not match then restore it
+            elif value != self._machine_saved_data[setting]:
+                self.machine.write_dollar_setting(setting, self._machine_saved_data[setting])
+                log('Restored {} from file: {}'.format(descriptions[setting], self._machine_saved_data[setting]))
+        # After startup, changes are intentional, so just overwrite internal value
+        elif value != self._machine_saved_data[setting]:
```

Comment:
don't do the updating in the callback! If the value changed unintentionally (which is the current assumption why it happens) we would just save it to the file.
Please create a separate function like "update_setting_51(value)" that is called from the maintenance screen when the user intentionally changes the selected spindle.